backend/fbdev/output.c

```diff
@@ -0,0 +1,301 @@
+	}
+	return true;
+}
+
+static bool output_test(struct wlr_output *wlr_output) {
+	if (wlr_output->pending.committed & WLR_OUTPUT_STATE_ENABLED) {
+		wlr_log(WLR_DEBUG, "Cannot disable a fbdev output");
+		return false;
+	}
+
+	if (wlr_output->pending.committed & WLR_OUTPUT_STATE_MODE) {
+		assert(wlr_output->pending.mode_type == WLR_OUTPUT_STATE_MODE_CUSTOM);
+	}
+
+	return true;
+}
+
+static bool output_commit(struct wlr_output *wlr_output) {
+	struct wlr_fbdev_output *output =
+		fbdev_output_from_output(wlr_output);
+
+	if (!output_test(wlr_output)) {
+		return false;
+	}
+
+	if (wlr_output->pending.committed & WLR_OUTPUT_STATE_MODE) {
+		if (!output_set_custom_mode(wlr_output,
+				wlr_output->pending.custom_mode.width,
+				wlr_output->pending.custom_mode.height,
+				wlr_output->pending.custom_mode.refresh)) {
+			return false;
+		}
+	}
+
+	if (wlr_output->pending.committed & WLR_OUTPUT_STATE_BUFFER) {
+		pixman_box32_t box = wlr_output->pending.damage.extents;
+		int32_t box_width = box.x2 - box.x1;
+
+		if (box_width) {
+			GLint format;
+			GLint type = GL_UNSIGNED_BYTE;
+			glGetIntegerv(GL_IMPLEMENTATION_COLOR_READ_FORMAT, &format);
+			struct fb_var_screeninfo scr_var = output->scr_var;
+			struct fb_fix_screeninfo scr_fix = output->scr_fix;
+			unsigned bytes_per_pixel = (scr_var.bits_per_pixel + 7) / 8;
+
+			wlr_log(WLR_DEBUG, "Updating box at %" PRIu32 "x%" PRIu32
+				" size %" PRIu32 "x%" PRIu32, box.x1, box.y1,
+				box_width, box.y2 - box.y1);
+			for (int y = box.y1; y < box.y2; y++) {
+				glReadPixels(box.x1, scr_var.yres - y - 1,
+					box_width, 1, format, type,
+					output->fbmem
+					+ (scr_var.yoffset + y) * scr_fix.line_length
+					+ (scr_var.xoffset + box.x1) * bytes_per_pixel);
+			}
+		}
+
+		glBindFramebuffer(GL_FRAMEBUFFER, 0);
+		wlr_egl_unset_current(output->backend->egl);
+		wlr_output_send_present(wlr_output, NULL);
+	}
+
+	return true;
+}
+
+static void output_rollback_render(struct wlr_output *wlr_output) {
+	struct wlr_fbdev_output *output =
+		fbdev_output_from_output(wlr_output);
+	assert(wlr_egl_is_current(output->backend->egl));
+	glBindFramebuffer(GL_FRAMEBUFFER, 0);
+	wlr_egl_unset_current(output->backend->egl);
+}
+
+static void output_destroy(struct wlr_output *wlr_output) {
+	struct wlr_fbdev_output *output =
+		fbdev_output_from_output(wlr_output);
+	wl_list_remove(&output->link);
+	wl_event_source_remove(output->frame_timer);
+	destroy_fbo(output);
+	free(output);
+}
+
+static const struct wlr_output_impl output_impl = {
+	.destroy = output_destroy,
+	.attach_render = output_attach_render,
+	.commit = output_commit,
+	.rollback_render = output_rollback_render,
+};
+
+bool wlr_output_is_fbdev(struct wlr_output *wlr_output) {
+	return wlr_output->impl == &output_impl;
+}
+
+static int signal_frame(void *data) {
+	struct wlr_fbdev_output *output = data;
+	wlr_output_send_frame(&output->wlr_output);
+	wl_event_source_timer_update(output->frame_timer, output->frame_delay);
+	return 0;
```

Comment:
Regarding this block:
```c
	output->committed = true;
	if (!output->timer_running) {
		wl_event_source_timer_update(output->frame_timer, output->frame_delay);
		output->timer_running = true;
	}
```

First I put it in the `if (box_width)` condition, because the outer condition `if (wlr_output->pending.committed & WLR_OUTPUT_STATE_BUFFER) { ` gets called often with a request to draw an empty damaged area (box_width = 0). However, when I put the snippet above inside the `if (box_width)` condition, wlroots will only draw one frame and then stop. Do you happen to know why that's the case?

If I put it inside the outer `if (wlr_output->pending.committed & WLR_OUTPUT_STATE_BUFFER) { ` condition, wlroots continues to draw frames, but is having a high CPU load again even if nothing changes on the screen.

Thanks a lot for all the help, @kennylevinsen!

Reply:
Okay, so I dug a bit further into it, and I think I managed to work around the issue:

```c
static bool output_attach_render(struct wlr_output *wlr_output,
		int *buffer_age) {
	...
	if (buffer_age != NULL) {
		*buffer_age = 1; // HACK HACK HACK
	}
	return true;
}

static bool output_commit(struct wlr_output *wlr_output) {
	struct wlr_fbdev_output *output =
		fbdev_output_from_output(wlr_output);

	if (!output_test(wlr_output)) {
		return false;
	}

	if (!output->backend->session->active) {
		return false;
	}

	...
	if (wlr_output->pending.committed & WLR_OUTPUT_STATE_BUFFER) {
		...
		wl_event_source_timer_update(output->frame_timer, output->frame_delay);
	}

	return true;
}

static int signal_frame(void *data) {
	struct wlr_fbdev_output *output = data;
	wlr_output_send_frame(&output->wlr_output);
	return 0;
}
```

For testing, I did a quicker implementation of the timer change: Set the timer in output_commit and leave it at that. If commit was called in quick succession, this might mess with the timing, but oh well.

The main issue, however, was the buffer_age: See here in `wlr_output_damage_attach_render`: https://github.com/swaywm/wlroots/blob/87836dcb55e15cd1f19f5549d4fcd666135dba15/types/wlr_output_damage.c#L158.

`buffer_age == 0` is handled the same was as too old a buffer, by requesting re-render and thus another commit. This leads to commits ad nauseam. I haven't looked into what the correct fix is here, but that's at least the cause.

Seems to render correctly in some simple tests, and does so with lower CPU utilization.

Reply:
I'll definitively give this a try over the next days. Thanks again, @kennylevinsen!

Reply:
This workaround works great, the CPU idles at 0% now when doing almost nothing (e.g. rewriting the clock every second in sway) now, instead of ~80%!